handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1359,7 +1359,7 @@ private int unwrap(ChannelHandlerContext ctx, ByteBuf packet, int length) throws
                 // Dispatch decoded data after we have notified of handshake success. If this method has been invoked
                 // in a re-entry fashion we execute a task on the executor queue to process after the stack unwinds
                 // to preserve order of events.
-                if (decodeOut.isReadable()) {
+                if (decodeOut.isReadable() && handshakePromise.isDone()) {
```

Comment:
this may result in data being dropped, which is observed in the `SocketSslEchoTest`. I ran your tests against https://github.com/netty/netty/pull/11210 and this seems to fix the issue. can you verify this PR fixes the root cause and if so I would propose we merge https://github.com/netty/netty/pull/11210, then merge the unit tests from this PR.

Reply:
@Scottmitch Thanks a lot, with #11210 I do not see the issue anymore!

Reply:
@violetagg I just merged #11210 ... Can you please revert the change in this class and rebase ? This way we can pull in your unit test and also verify it passes on the CI.

Reply:
done